ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/Committer.scala

```diff
@@ -78,46 +87,63 @@ private[committer] trait Committer[PartialResult] extends SubmissionExecutor {
       logEntry -> ctx.getOutputs.toMap
     }
 
-  def dryRun(
+  def runWithPreExecution(
       submission: DamlSubmission,
       participantId: ParticipantId,
       inputState: DamlStateMapWithFingerprints,
-  ): PreexecutionResult = {
-    runTimer.time { () =>
-      // TODO(miklos): Create context for pre-execution here.
+  ): PreExecutionResult = {
+    preExecutionRunTimer.time { () =>
       val commitContext = new CommitContext {
         override def getRecordTime: Option[Time.Timestamp] = None
 
         override def getParticipantId: ParticipantId = participantId
 
         override def inputsWithFingerprints: DamlStateMapWithFingerprints = inputState
       }
-      preexecute(submission, participantId, inputState, commitContext)
+      preExecute(submission, participantId, inputState, commitContext)
     }
   }
 
-  private[committer] def preexecute(
+  private[committer] def preExecute(
       submission: DamlSubmission,
       participantId: ParticipantId,
       inputState: DamlStateMapWithFingerprints,
       commitContext: CommitContext,
-  ): PreexecutionResult = {
+  ): PreExecutionResult = {
     val logEntry = runSteps(commitContext, submission)
-    PreexecutionResult(
+    PreExecutionResult(
       readSet = commitContext.getAccessedInputKeysWithFingerprints.toMap,
       successfulLogEntry = logEntry,
       stateUpdates = commitContext.getOutputs.toMap,
-      // TODO(miklos): Take out-of-time-bounds log entry and min/max record time from context.
       outOfTimeBoundsLogEntry = constructOutOfTimeBoundsLogEntry(commitContext),
-      minimumRecordTime = Timestamp.MinValue,
-      maximumRecordTime = Timestamp.MaxValue,
+      minimumRecordTime = commitContext.minimumRecordTime
+        .map(Timestamp.assertFromInstant)
+        .getOrElse(Timestamp.MinValue),
```

Comment:
This default seems to me safe only for some submission types though, so I suggest specializing.

Reply:
We could make `minimumRecordTime` optional, however, the check that needs to be done during post-execution should be a simple comparison (i.e., whether the record time at that time is within limits). Not making this field optional simplifies the post-execution code as any record time other than `Timestamp.MinValue` should be larger than the default minimum record time set here.

Reply:
Pre-executing a transaction submission and not getting back minimum/maximum record time is not a valid state AFAIU, so I think we shouldn't set the widest possible window nor avoid setting it but we should rather throw. This is committer-specific though, so I think the committer should tell what to do in such cases.

Reply:
E.g. an option could be specializing the context, another one could be to have the committer validate the context before it is used to return the pre-execution result.

Reply:
I'm not sure if it's worth introducing more complexity around these checks (i.e., make the left side of the interval optional while the right side is mandatory). As we are controlling the full flow I'm leaning towards simplicity and just have a very relaxed left side of the interval that needs to be checked (i.e., any record time will adhere to it). What disadvantage can you see with keeping the implementation as-is? We have unit tests covering what is expected from the `Committer` implementations plus the default values.

Reply:
I think we shouldn't fill here missing committer-specific data with general defaults that are not generally safe, it is confusing and we might miss something it in the future. You could also just throw instead and expect the specialized committers to always set a value (possibly a default one).

Reply:
Why are you suggesting that `Timestamp.MinValue` as the default for a non-existant left interval point is not safe? I'm happy to make both of these record times optional, i.e., we will send these as part of a proto3 message in which every field is optional anyway.

Reply:
It's not just about minimum but both extremes and the reason is that minimum and maximum record time _must_ be filled in by the transaction and configuration committers during pre-execution. The matter here is the following: as a reader, I don't understand what assumption is made about the concrete committers, i.e. I read "during pre-execution, committers will fill min and max record time and if they don't then it means we should take the biggest possible window (independently of which committers they are)", which is not always true. I'd prefer reading "during pre-execution, committers will fill min and max record time (possibly with defaults), else it's a bug".

Reply:
Makes sense -- we can have an optional here then for min/max record time.